report.go

```diff
@@ -0,0 +1,212 @@
+package gomoss
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/anaskhan96/soup"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Region specifies similarity from line a to line b
+type Region struct {
+	from int64
+	to   int64
+}
+
+// Source is structure that containing plagiarism result of a source code
+type Source struct {
+	Filename   string
+	Similarity float64
+	Regions    []Region
+}
+
+// Match means compare the plagiarism between two source codes
+type Match struct {
+	Left, Right Source
+}
+
+// Report is data extracted data from Moss URL
+type Report struct {
+	Matches []Match
+}
+
+var (
+	ErrInvalidRegion = errors.New("Invalid region")
+)
+
+func mustParse(url *url.URL, rawurl string) *url.URL {
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		panic(err)
+	}
+	return u
+}
+
+// FindLinks find all analysis related hrefs in HTML
+func findLinks(ref *url.URL) []string {
+	resp, _ := soup.Get(ref.String())
+
+	doc := soup.HTMLParse(resp)
+	anchors := doc.FindAll("a")
+	exist := make(map[string]bool)
+	var links []string
+	for _, anchor := range anchors {
+		link := anchor.Attrs()["href"]
+		if strings.Contains(link, "match") && !exist[link] {
+			exist[link] = true
+			links = append(links, link)
+		}
+	}
+	return links
+}
+
+// DownloadURL downloads URL to file
+func DownloadURL(ref *url.URL, filename string) error {
+	resp, err := soup.Get(ref.String())
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(filename)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+
+	reg := regexp.MustCompile(`http:\/\/moss.stanford.edu\/results\/\d*\/`)
+	resp = reg.ReplaceAllString(resp, "")
+
+	_, err = f.Write([]byte(resp))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Download downloads all HTML files from given URL recursively
+func Download(ctx context.Context, ref *url.URL) (string, error) {
+	tempdir, err := ioutil.TempDir(os.TempDir(), "temp")
```

Comment:
nit: you shouldn't leave temporary directories around :( every time you use `ioutil.TempDir` it _must_ be followed by `os.RemoveAll(tempdir)`. That means that `Download()` should receive the directory where it will store the files.